addon/-private/system/relationships/state/belongs-to.js

```diff
@@ -31,11 +31,14 @@ BelongsToRelationship.prototype.setRecord = function(newRecord) {
 };
 
 BelongsToRelationship.prototype.setCanonicalRecord = function(newRecord) {
+  if (this.canonicalState) {
+    this.removeCanonicalRecord(this.canonicalState);
+  }
```

Comment:
I think the real fix is the fact that `removeCanonicalRecord` ends up calling `this.flushCanonicalLater();` somewhere in its call stack. Instead of moving this check up I'd prefer to see it left where it is an add an explicit call to `this.flushCanonicalLater();` before `this.setHasData(true);`.